File: Programming Languages/CPP/0020_Valid_Parentheses/solution_3.cpp
// 0020. Valid Parentheses
// Difficulty : Easy
// Tags       : string, stack
// Solved on  : 2020-12-17
// Attempt    : #3
// LeetCode   : https://leetcode.com/problems/valid-parentheses/

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(char c : s)
        {
            if(c=='(' || c=='{' || c=='[')
            {
                st.push(c);
            }
            else if(c==')' || c=='}' || c==']')
            {   if(st.empty())
                    return false;
                char ch=st.top();
                if(c==')'){
                    if(ch!='(') return false; 
                }
                else if(c=='}')
                {
                    if(ch!='{') return false;
                }
                else if(c==']')
                {
                    if(ch!='[') return false;
                }
                st.pop();
            }
            
            
           
        }
        if(st.empty()) return true;
        return false;
    }
};

File: DSA/Math/0062_Unique_Paths/python3/solution_3.py
# 0062. Unique Paths
# Difficulty : Medium
# Tags       : math, dynamic-programming, combinatorics
# Solved on  : 2026-04-30
# Attempt    : #3
# LeetCode   : https://leetcode.com/problems/unique-paths/

class Solution:
    def __init__(self):
        self.dp = [ [ -1 for _ in range(101)] for _ in range(101)]
    
    def uniquePaths(self, m: int, n: int) -> int:
        return self.upath(0,0,m,n);
        
    def upath(self,i:int,j:int,m:int,n:int):
        if i >=m or j >=n:
            return 0
        if self.dp[i][j] != -1:
            return self.dp[i][j]
            
        if i == m-1 and j == n-1:
            return 1
        self.dp[i][j] = self.upath(i+1,j,m,n) + self.upath(i,j+1,m,n)
        return self.dp[i][j]